manpage/utitool.1
.TH UTITOOL "1" "October 2026" "utitool 2.0.0"

.SH NAME
utitool \- Retrieve file and system Uniform Type Identifier (UTI) information.

.SH SYNOPSIS
.P
\fButitool\fR [\fB\-e\fR file extension]
.P
\fButitool\fR [\fB\-u\fR UTI]
.P
\fButitool\fR [\fB\-l\fR] [\fB\-j\fR]
.P
\fButitool\fR [\fB\-a\fR] [\fB\-j\fR]
.P
\fButitool\fR [\fB\-m\fR] path...
.P
              [\fB\-h\fR] [\fB\--version\fR]

.SH DESCRIPTION
.P
\fButitool\fR can retrieve UTI information for one or more files. Separately, it can be used to retrieve information about a specific UTI (eg. \fInet.daringfireball.markdown\fR) or a specific file extension (eg. \fImd\fR).

A third mode retrieves and lists system-held UTI information or UTI-relevant app information.

.SH USAGE

Options are used in specific modes.

\fBFile Mode\fR
.P
{path}                            The path to the target file or files.
.P
\fB\-m\fR | \fB\-\-more\fR                       Display extra information about each file.

\fBSpecific UTI or File Extension Mode\fR
.P
\fB\-u\fR | \fB\-\-uti\fR {UTI}                  The UTI you are interested in.
.P
\fB\-e\fR | \fB\-\-extension\fR {extension}      The file extension you are interested in.

\fBSystem Information Mode\fR
.P
\fB\-l\fR | \fB\-\-list\fR                       List information on the UTIs recorded by the system.
.P
\fB\-a\fR | \fB\-\-apps\fR                       List information on the apps recorded by the system.
.P
\fB\-j\fR | \fB\-\-json\fR                       Output the above information as JSON.

\fBSet Mode\fR
.P
\fB\-s\fR | \fB\-\-set\fR {UTI} {app}            Specify a new default app to open files of type {UTI}.

\fBAll Modes\fR
.P
\fB\-h\fR | \fB\-\-help\fR                       The help screen.
.P
\fB\-\-version\fR                         Show \fButitool\fR version information.
.PP

.SH EXAMPLES
.P
utitool ~/Downloads/files
.P
utitool \-\-more ~/Downloads/files/file.md
.P
utitool \-\-uti net.daringfireball.markdown
.P
utitool \-\-extension pdf
.P
utitool \-\-list
.P
uitool \-\-list \-\-json
.P
utitool \-lj
.P
uitool \-\-apps \-\-json

.SH COPYRIGHT
Code and design copyright © 2026, Tony Smith (\fB@smittytone\fR). Source code available under the MIT licence.

.SH "SEE ALSO"
\fIhttps://smittytone.net/utitool/index.html\fR


CHANGELOG.md
## Release Notes ##

- 2.0.0 *Unreleased*
    - **Breaking** Minimum supported macOS is now 12.4.
    - Update to `Clicore 0.7.0` (tracks `develop`).
    - Project build structure changes.
- 1.2.2 *21 August 2025*
    - Add `man` page.
    - Use `Clicore` for common CLI and Stdio functionality.
- 1.2.1 *17 August 2025*
    - Support combined command line switches, eg. `-lj` (list UTIs, output to JSON).
    - Under-the-hood non-functional changes: namespace more files using structs.
- 1.2.0 *12 August 2025*
    - Add `--list` and `--app` options to retrieve and display system-recorded information for, respectively, UTIs and apps.
    - Support machine- or human-readable output for the above commands, with the `--json` switch for the former.
    - Add `UTITOOL_USE_DARK_COLOUR` environment variable for users with dark-on-light Terminal colouring.
    - Check for correct macOS version.
    - Reorganise the code base to separate out key components.
- 1.1.0 *1 August 2025*
    - Add extra file information.
    - Add `--extension` option to show info about specified a file extension.
    - Add `--uti` option to show info about a specified UTI.
    - Add `--more` option to show extra UTI information about specified files.
- 1.0.4 *23 July 2021*
    - Add async signal safe ctrl-c trapping code.
- 1.0.3 *15 June 2021*
    - Add support for macOS 11 Big Sur’s `UTType` API.
- 1.0.2 *8 February 2021*
    - Minor change to help text.
- 1.0.1 *4 February 2021*
    - Tiny bit of refactoring.
- 1.0.0 *12 January 2021*
    - Initial public release.


utitool/utitool.swift
/*
    utitool
    utitool.swift

    Copyright © 2026 Tony Smith. All rights reserved.

    MIT License
    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in all
    copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
    SOFTWARE.
*/

import Foundation
import Clicore


/**
    Main entry point
 */
@main
struct Utitool {

    public static func main() async throws {

        // Set up Ctrl-C handling
        Stdio.enableCtrlHandler("utitool interrupted -- halting")

        // FROM 2.0.0
        var settings = Settings()

#if os(macOS)
        // Use emoji markers on macOS
        Stdio.settings.useEmoji = true
#endif

        // FROM 1.2.0
        // Check for a colour shift
        if let _ = ProcessInfo.processInfo.environment["UTITOOL_USE_DARK_COLOUR"] {
            settings.highlightColour = String(Stdio.ShellColour.blue)
        }

        // Get the command line args...
        let collatedArguments = Cli.unify(args: CommandLine.arguments)

        // ...and process them
        if collatedArguments.count == 0 {
            // No user args? Just show the help info
            showHelp()
        } else {
            // Process the (separated) arguments
            var previousArgument = ""
            var requiresValue = -1
            for argument in collatedArguments {
                if requiresValue > 0 {
                    // Make sure we're not reading in an option rather than a value
                    if argument.prefix(1) == "-" {
                        Stdio.reportErrorAndExit("Missing value for \(previousArgument)")
                    }

                    switch requiresValue {
                        case 1:
                            getExtensionData(argument, settings)
                            Stdio.exitApp()
                        case 2:
                            getUtiData(argument, true, settings)
                            Stdio.exitApp()
                        default:
                            break
                    }

                    requiresValue = -1
                    continue
                }

                switch argument {
                    case "--extension", "-e":
                        requiresValue = 1
                        previousArgument = argument
                    case "--uti", "-u":
                        requiresValue = 2
                        previousArgument = argument
                    case "--more", "-m":
                        settings.showMoreInfo = true
                    case "--list", "-l":
                        settings.doLaunchServicesReadUtis = true
                    case "--apps", "-a":
                        settings.doLaunchServicesReadApps = true
                    case "--json", "-j":
                        settings.doOutputJson = true
                    case "--set", "-s":
                        settings.doSetDefaultApp = true
                    case "-h", "-help", "--help":
                        showHelp()
                        Stdio.exitApp()
                    case "--version":
                        showHeader()
                        Stdio.exitApp()
                    default:
                        if argument.prefix(1) == "-" {
                            Stdio.reportErrorAndExit("Unknown argument: \(argument)")
                        } else {
                            settings.files.append(argument)
                        }
                }

                // Trap commands that come last and therefore have missing args
                if requiresValue > 0 && argument == collatedArguments.last {
                    Stdio.reportErrorAndExit("Missing value for argument \(argument)")
                }
            }

            // FROM 2.0.0
            // Set a default app
            if settings.doSetDefaultApp {
                if settings.files.count != 2 {
                    Stdio.reportErrorAndExit("Missing the UTI and/or default app")
                }

                switch await setDefaultApp(settings) {
                    case .failure(let error):
                        Stdio.reportErrorAndExit("\(error.localizedDescription)")
                    case .success(let message):
                        Stdio.report(message)
                        Stdio.exitApp()
                }
            }

            // Run initial LSServices functions
            if settings.doLaunchServicesReadUtis && settings.doLaunchServicesReadApps {
                // There can be only one...
                settings.doLaunchServicesReadApps = false
            }

            if settings.doLaunchServicesReadApps {
                readLaunchServicesRegister(true, settings)
                Stdio.exitApp()
            }

            if settings.doLaunchServicesReadUtis {
                readLaunchServicesRegister(false, settings)
                Stdio.exitApp()
            }

            // Convert passed paths to URLs for processing
            var count = 0
            if settings.files.count > 0 {
                for file in settings.files {
                    let path = Path.getFullPath(file)
                    var isDir: ObjCBool = false

                    // Check that we're only dealing with files
                    if FileManager.default.fileExists(atPath: path, isDirectory: &isDir) {
                        if isDir.boolValue {
                            // Ignore directories
                            continue
                        }

                        // Make a URL from the path
                        let url = URL(fileURLWithPath: path, isDirectory: false)

                        // And output the UTI if we can
                        if let uti = url.typeIdentifier {
                            var extra = ""
                            if url.known {
                                extra = "UTI is registered with the system"
                            } else if uti.hasPrefix("dyn") {
                                extra = "UTI was dynamically assigned"
                            }

                            if settings.showMoreInfo {
                                Stdio.report("UTI for \(settings.highlightColour)\(path)\(String(.normal)) is \(settings.highlightColour)\(uti)\(String(.normal))")
                                getUtiData(uti, false, settings)
                            } else {
                                Stdio.report("UTI for \(settings.highlightColour)\(path)\(String(.normal)) is \(settings.highlightColour)\(uti)\(String(.normal)) (\(extra))")
                            }
                        } else {
                            Stdio.reportError("Could not get UTI for \(path)")
                        }

                        // Tally the number of files reported on
                        count += 1
                    } else {
                        Stdio.reportError("\(path) is not a valid file reference")
                    }
                }
            } else if !settings.doLaunchServicesReadApps && !settings.doLaunchServicesReadUtis {
                // No reported files? Issue a warning
                Stdio.report("No files specified or present")
            }
        }

        // Exit gracefully
        Stdio.exitApp()
    }
}
